src/compass/utils/runconfig.py

```diff
@@ -285,3 +333,27 @@ def gpu_enabled(self):
     @property
     def gpu_id(self):
         return self.groups.worker.gpu_id
+
+    def as_dict(self):
+        # convert to dict first then dump to yaml
+        self_as_dict = {}
+        for key, val in self.__dict__.items():
+            if key == 'groups':
+                val = unwrap_to_dict(val)
+            elif key == 'bursts':
+                # just date in datetime obj as string
+                date_str = lambda b : str(b.sensing_start.date()).split()[0]
```

Comment:
```suggestion
                date_str = lambda b : b.sensing_start.date().strftime("%Y%m%d")
```
I was just trying to load the metadat "metadata.json" and did a quick experiment to convert the dict to namespace using 

```
import json
from compass.utils.wrap_namespace import wrap_namespace

fid = open("metadata.json","r")
m = json.load(fid)
s = wrap_namespace(m)
```
and then I can say:

```
s.bursts
```
but it fails for 

```
 s.bursts.t173_iw2_b863_2017-05-10_VH
  File "<ipython-input-6-e3d644dcbbbd>", line 1
    s.bursts.t173_iw2_b863_2017-05-10_VH
                                 ^
SyntaxError: leading zeros in decimal integer literals are not permitted; use an 0o prefix for octal integers
```
And that is because Python packages and modules can not use hyphens, only underscores. 

If you go with my suggestion above, then we can say:

```
In [11]: s.bursts.t173_iw2_b863_20170510_VV.sensing_start
Out[11]: '2017-05-10 13:43:38.353501'```

which is cool! i.e., we can access the metadata as the members of a python object.

Reply:
For completeness I think we should only write the metadata that are relevent to the output geocoded burst and avoid writing those that are specifically pointing to the input burst. After moving this part to within ```run```, let's create a new dict let's say called ```gslc_dict``` which has the following keys:

dict_keys(['sensing_start', 'radar_center_frequency', 'wavelength', 'azimuth_steer_rate', 'azimuth_time_interval', 'slant_range_time', 'starting_range', 'range_sampling_rate', 'range_pixel_spacing', 'azimuth_fm_rate', 'doppler', 'range_bandwidth', 'polarization', 'burst_id', 'platform_id', 'center', 'border', 'orbit', 'tiff_path', 'i_burst', 'range_window_type', 'range_window_coefficient'])

which we can borrow from ```runconfig['bursts'][burst.burst_id]```

Let's add sensing_stop (you need to take care of the datetime object)
gslc_dict['sensing_stop'] =  gslc_dict['sensing_start'] + radargrid.length *  azimuth_time_interval

and then let's add the geogrid
gslc_dict['geogrid'] = runconfig['geogrids'][burst.burst_id]  

and add run config
gslc_dict["runconfig"] = runconfig['groups']

add the nodata value:
gslc_dict['nodata'] = NO_DATA_VALUE

we should read this with the reader
gslc_dict['input_data_ipf_version'] =  

empty for now but would be nice to keep track of it. 
gslc_dict['compass-isce3-version'] = ""



And then let's write gslc_dict to file (json, yaml). That way when we load it back as:
```
fid = open("metadata.json","r")
m = json.load(fid)
s = wrap_namespace(m)
```
this time we can easily access the metadata at the top level of the object as for example:

```
s.sensing_start
```
in opposed to what we had above as

```
s.bursts.t173_iw2_b863_20170510_VV.sensing_start
```
You may make sure that s.polarization represents all polarizations in the output geocoded burst. 

This should be most of the light metadata. @vbrancat @yunjunz  can you think of other light metadata that may be useful to add.  I can imagine that we will add bunch of metadata about correction when we get there.

Reply:
@LiangJYu and @vbrancat sorry this is becoming a bit long comment. Happy to talk offline if needed. What do you think about this idea to define a class for geocoded burst slc which has reader and writer methods. Here is a very general idea to build on:

```
class GSLC:
    def __init__(self):
    self.sensing_start = None
    self.sensing_stop = None
    self.polarization = None
    self.orbit = None
    # ....

    def as_dict(self):
        return gslc_dict

    def sensing_start(self, s):
        self.sensing_start = s

    def sensing_stop(self, s):
        self.sensing_stop = s

    # and setters for other parameters ...

    def load_from_dict(self, input_dict):
        # We can assume input dict is the runconfig Then for max clarity we could
        # go one by one and set all data memebers

    def write_metadata(self, metadata_file):
        self_as_dict = self.as_dict()
        with open(metadata_file, 'w') as f_json:
            json.dump(self_as_dict, f_json)

    def load_metadata(self, metadata_file):
        fid = open("metadata.json","r")
        meta_dict = json.load(fid)
        self = wrap_namespace(meta_dict)
```
I think this will significantly increase the clarity and will make the maintenance of the product and its metadata much easier. It is much more clear to see each of the metadata (class data members) and how they are set before writing them to file. Also with a simple method we should be able to reconstruct the object from file.

Reply:
I like the idea of the class and it can simplify a lot the product reader and writer (which is on our TO DO list anyway). Plus, the class can be used in other parts of the code (e.g. stitching). @LiangJYu will take care of the metadata part

Reply:
For the metadata, the following could also be useful: 'length', 'width', 'earth_radius', 'satellite_height', 'orbit_direction', 'antenna_side', 'pulse_repetition_frequency' and the EPSG code.

Reply:
@yunjunz length, width and epsg are must to have and included in geogrid mentioned above. what do you mean by antenna_side? Do you mean look_direction? If yes I prefer look_direction compared to antenna_side.

The orbit is included which is the satellite XYZ and velocity. I'm not sure if we should provide satellite_height and earth_radius. satellite_height if needed , they can be derived from the orbit info. One can convert XYZ to llh, look up DEM height at ll and subtract from h of llh to get to satellite_height. 

It would be good to have orbit_direction in the metadata.

Reply:
@hfattahi yes, for the antenna_side, I meant "look_direction", that sounds good to me too. 

The calculation of satellite_height and the used earth_radius is not very end user friendly, I think it will be useful to have them unless this causes confusion on our side.

Reply:
@hfattahi @yunjunz `look direction` easier to add to s1_reader/s1_burst. Will simply returning `right` be sufficient?

`orbit_direction` will also be easier to add to s1_reader/s1_burst.

Reply:
I think look direction returning from s1 reader is fine. And yes look direction for S1 is fixed to "right". 
And yes orbit direction can come with the reader.